### DISCUSSION TITLE
Examples for use with nonlinear optimization solvers

### DISCUSSION THREAD

**Discussion by sandeep026** [aesara-devs/aesara#1521] [Q&A]
Hi there,

I am new to Aesera here. I work on solving optimal control problems numerically. It deals with solving constrained nonlinear optimization problems. Like any optimization problem, solutions can be effiiently computed if the derivatives like hessian, jacobian and gradient are computed.  At present, I use casadi where I create symbolic primitives and create functions for the optimizaiton problem and use the auto diff capabilities to compute the derivatives and solve the optimization problem. I consider this aspect similar to what aesara offers. 
Before I start experimenting with aesara I have the following doubts.

1. Does aesera use automatic differentiation/symbolic differentiation? 
2. Are there any benchmarks with other libraries? 
3. Has it been used for solving optimal control problems?
4. Does it support dense and sparse problems (dense/sparse hessian etc.)?
5. Will the project have longterm development and support?

**Comment by brandonwillard** [comment]
Thanks for the interest in Aesara!

> * Does aesera use automatic differentiation/symbolic differentiation?

It uses AD.

> * Are there any benchmarks with other libraries?

Exactly which kinds of benchmarks are you interested in?  This project is a fork of Theano, and there may be some older relevant benchmarks, but we haven't made direct comparisons with other modern tensor libraries because this one was intended primarily to serve as a wrapper around others (e.g. JAX) for the purpose of facilitating symbolic computations.

> * Has it been used for solving optimal control problems?

I'm not presently aware of any such projects explicitly using Aesara.

> * Does it support dense and sparse problems (dense/sparse hessian etc.)?

Sparse matrices are supported and can be used with the Hessian-computing operators, but some work may need to be done in order to guarantee that sparsity is maintained all throughout the resulting computations.  Luckily, it's fairly straightforward to see exactly which operations would take place and the sparsity of the intermediate results between them.

> * Will the project have longterm development and support?

The maintainers of this project are currently involved full-time in [a new endeavor in the LLM space](https://dottxt.co/), so support will be very limited.

Likewise, Aesara is not a funded open-source project, so everything is completely volunteer.  If you need true longevity, consistently good average performance, and support for all standard tensor library features, we can really only recommend investing in a library like PyTorch.

If you have specific requirements that Aesara meets and you would like to know where else to look for those, feel free to ask.
